sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,745 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+extern crate serde;
+extern crate serde_hjson;
+
+use anyhow::{bail, Result};
+use memoffset::offset_of;
+use serde::Deserialize;
+
+use std::fmt;
+use std::fs;
+use std::io::{self, Read};
+use std::ops::Range;
+use std::path::Path;
+
+use std::{mem, ptr};
+
+use crate::crypto::rsa::RsaPrivateKey;
+use crate::crypto::sha256;
+use crate::util::parse_int::ParseInt;
+
+// RV32 NOP (addi, x0, x0, 0)
+const RV32_NOP: u32 = 0x00000013;
+
+// RSA-3072 signature length in bytes
+const RSA_SIGNATURE_LEN_BYTES: usize = 3072 / 8;
+
+// At most 32 match descriptors per patch
+const N_MATCH_DESCRIPTORS: usize = 32;
+
+// Patch match table size
+const PATCH_MATCH_TABLE_SIZE: usize =
+    N_MATCH_DESCRIPTORS * mem::size_of::<RomPatchMatchDescriptor>();
+
+// The code section offset from the patch start
+const PATCH_CODE_SECTION_OFFSET: usize = mem::size_of::<RomPatchHeader>() + PATCH_MATCH_TABLE_SIZE;
+
+#[allow(dead_code)]
+enum MultiBitBool {
+    Bool4True = 0x6,
+    Bool4False = 0x9,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchSubRoutineHjson {
+    base_addr: String,
+    code: Vec<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchRegionHjson {
+    match_addr: String,
+    locked: bool,
+    patch_code: Vec<String>,
+    sub_routines: Option<Vec<RomPatchSubRoutineHjson>>,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchHjson {
+    revision: u8,
+    base_addr: String,
+    regions: Vec<RomPatchRegionHjson>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct RomPatchPartitionHjson {
+    patches: Vec<RomPatchHjson>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
+#[repr(packed)]
+struct RomPatchMatchDescriptor {
+    match_base: u32,
+    remap_base: u32,
+}
+
+impl RomPatchMatchDescriptor {
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = u32::from_str(&region.match_addr)?;
+
+        let size_dwords = (region.patch_code.len() as u32) & 0xf;
+        let locked = if region.locked { 1 << 31 } else { 0 };
+
+        let match_address = match_addr & 0x7ffffff;
+        let match_base = locked | (size_dwords << 27) | match_address;
+
+        Ok(RomPatchMatchDescriptor {
+            match_base,
+            remap_base,
+        })
+    }
+
+    fn match_addr(&self) -> u32 {
+        self.match_base & 0xffff
+    }
+
+    fn remap_addr(&self) -> u32 {
+        self.remap_base
+    }
+
+    fn locked(&self) -> bool {
+        (self.match_base >> 31) != 0
+    }
+
+    fn size(&self) -> u8 {
```

Comment:
I suggest renaming `size` to something more precise, like `size_dwords` (which I think is what it is here?)

Reply:
Not relevant anymore with the updated PR.